src/vm/types/signatures.rs

```diff
@@ -215,6 +215,14 @@ impl ListTypeData {
         (*self.entry_type, self.max_len)
     }
 
+    // if checks like as-max-len pass, they may _reduce_
+    //   but should not increase the type signatures max length
+    pub fn reduce_max_len(&mut self, new_max_len: u32) {
+        if new_max_len <= self.max_len {
+            self.max_len = new_max_len;
+        }
```

Comment:
Can it ever be the case that the code calls `reduce_max_len()` with a `new_max_len` that is larger than `self.max_len`?

Reply:
Yep— if a list already had a small type annotation when passed to as-max-Len, new_max_len would be greater than the original. In that case, the list type annotation should remain as is.